File: FACULDADE/AEDS2 (Algoritmos e Estruturas de Dados 2)/Treino/Main.java
import java.util.Scanner;

//Treinando estrutura de arvore em java pra prova pratica

public class Main {
  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    Arvore arvore = new Arvore();
    while (scanner.hasNext()) {
      int input = scanner.nextInt();
      if (input == 0)
        break;
      System.out.println("raiz!" + arvore.raiz.elemento + " " + input);
      arvore.inserir(input);

    } /*
       * System.out.println("raiz: " + arvore.raiz.elemento);
       * System.out.println("dir: " + arvore.raiz.dir.elemento);
       * System.out.println("esq: " + arvore.raiz.esq.elemento);
       */
    arvore.imprimir();

    scanner.close();
  }

  public static class Arvore {
    No raiz;

    Arvore() {
      this.raiz = new No();
    }

    public static class No {
      int elemento;
      No esq, dir;

      No() {
        this.elemento = 0;
        this.esq = null;
        this.dir = null;
      }

      No(int x) {
        this.elemento = x;
        this.esq = null;
        this.dir = null;
      }

    }

    public void inserir(int elemento) {
      if (this.raiz == null || this.raiz.elemento == 0)
        this.raiz = new No(elemento);
      else {
        No percorre = this.raiz;
        while (percorre != null) {
          if (elemento < percorre.elemento) {
            if (percorre.esq == null) {
              percorre.esq = new No(elemento);
              percorre = null;
            } else {
              percorre = percorre.esq;
            }
          } else {
            if (percorre.dir == null) {
              percorre.dir = new No(elemento);
              percorre = null;
            } else {
              percorre = percorre.dir;
            }
          }

        }
      }
    }

    public void imprimir() {
      imprimir(this.raiz);
    }

    public void imprimir(No raiz) {
      if (raiz != null) {
        imprimir(raiz.esq);
        System.out.println(raiz.elemento);
        imprimir(raiz.dir);
      }
    }

  }
}

File: FACULDADE/AEDS2 (Algoritmos e Estruturas de Dados 2)/Treino/main.c
#include <stdio.h>
#include <stdlib.h>
//TREINANDO ARVORE EM C
// NO
typedef struct No {
  int elemento;
  struct No *esq, *dir;
} No;

// ARVORE
typedef struct {
  No *raiz;
} Arvore;

/* INSERIR com ponteiro duplono No raiz para poder modificar a raiz se a árvore
estiver vazia. */
void inserir(int elemento, No **raiz) {
  if (*raiz == NULL) {
    No *novoNo = (No *)malloc(sizeof(No));
    novoNo->elemento = elemento;
    novoNo->esq = NULL;
    novoNo->dir = NULL;
    *raiz = novoNo;
  } else {
    No *atual = *raiz;
    if (elemento < atual->elemento) { // esq
      inserir(elemento, &atual->esq);
    } else if (elemento > atual->elemento) { // dir
      inserir(elemento, &atual->dir);
    }
  }
}

// Caminhamento em ordem
void imprimir(No *raiz) {
  if (raiz != NULL) {
    imprimir(raiz->esq);
    printf("Caminhamento em ordem: %d\n", raiz->elemento);
    imprimir(raiz->dir);
  }
}

// Caminhamento pos ordem
void imprimirPos(No *raiz) {
  if (raiz != NULL) {
    imprimirPos(raiz->esq);
    imprimirPos(raiz->dir);
    printf("Caminhamento pos ordem: %d\n", raiz->elemento);
  }
}

// Caminhamento pos ordem
void imprimirPre(No *raiz) {
  if (raiz != NULL) {
    printf("Caminhamento pre ordem: %d\n", raiz->elemento);
    imprimirPre(raiz->esq);
    imprimirPre(raiz->dir);
  }
}

int main(void) {

  Arvore *tree = (Arvore *)malloc(sizeof(Arvore));
  tree->raiz = NULL; // inicializa raiz como NULL

  int x, d;
  scanf("%d", &x);

  for (int i = 0; i < x; i++) {
    scanf("%d", &d);
    inserir(d, &tree->raiz);
  }

  imprimirPre(tree->raiz);
  // libera memoria
  free(tree->raiz);
  free(tree);

  return 0;
}
